src/org/etools/j1939_84/controllers/part07/Part07Step02Controller.java

```diff
@@ -55,12 +62,54 @@
     protected void run() throws Throwable {
         // 6.7.2.1.a DS DM23 ([send Request (PGN 59904) for PGN 64949 (SPNs 1213-1215, 1706, and 3038)]) to each OBD
         // ECU.
-        // 6.7.2.2.a Fail if no OBD ECU reports previously active DTC.
-        // 6.7.2.2.b Fail if reported previously active DTC does not match DM12 active DTC from part 6.
-        // 6.7.2.2.c Fail if any ECU does not report MIL off and not flashing.
-        // 6.7.2.2.d Fail if NACK not received from OBD ECUs that did not provide a DM23 message.
-        // 6.7.2.3.a Warn if any ECU reports > 1 previously active DTC.
+        var dsResults = getDataRepository().getObdModuleAddresses()
+                                           .stream()
+                                           .map(address -> getDiagnosticMessageModule().requestDM23(getListener(),
+                                                                                                    address))
+                                           .collect(Collectors.toList());
+
+        List<DM23PreviouslyMILOnEmissionDTCPacket> packets = filterPackets(dsResults);
+
+        // 6.7.2.2.b Fail if no OBD ECU reports previously active DTC.
+        long activeDtcCount = packets.stream()
+                                     .filter(p -> getDataRepository().isObdModule(p.getSourceAddress()))
+                                     .filter(p -> !p.getDtcs().isEmpty())
+                                     .count();
+        if (activeDtcCount == 0) {
+            addFailure("6.7.2.2.b - No OBD ECU reported a previously active DTC");
+        }
+
+        packets.stream().peek(this::save).forEach(p -> {
+            // 6.7.2.2.b Fail if reported previously active DTC does not match DM12 active DTC from part 6.
+            List<DiagnosticTroubleCode> dm6Dtcs = getDataRepository().getObdModule(p.getSourceAddress())
+                                                                     .get(DM6PendingEmissionDTCPacket.class)
+                                                                     .getDtcs();
+            if (!p.getDtcs().equals(dm6Dtcs)) {
```

Comment:
you can't do this.  The `equals()` method on the `DiagnosticTroubleCode` includes conversion method and occurrence count.  The spec expects that we only compare the SPN and FMIs.

There is a help method on the `StepController' called `listContainsDTC()`

Reply:
I created a method call dtcListsAreSame in the StepController to handle this check because the requirements actually say that the dtc lists must match.